x-pack/plugin/profiler/src/test/java/org/elasticsearch/xpack/profiler/StackFrameTests.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.profiler;
+
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertToXContentEquivalent;
+
+public class StackFrameTests extends ESTestCase {
+    public void testCreateFromSource() {
+        // tag::noformat
+        StackFrame frame = StackFrame.fromSource(
+            Map.of("Stackframe", Map.of(
+                "file", Map.of("name", "Main.java"),
+                "function", Map.of(
+                        "name", "helloWorld",
+                        "offset", 31733
+                    ),
+                "line", Map.of("number", 22),
+                "source", Map.of("type", 3))
+                )
+        );
+        // end::noformat
+        assertEquals(Integer.valueOf(3), frame.sourceType);
+        assertEquals("Main.java", frame.fileName);
+        assertEquals("helloWorld", frame.functionName);
+        assertEquals(Integer.valueOf(31733), frame.functionOffset);
+        assertEquals(Integer.valueOf(22), frame.lineNumber);
+    }
+
```

Comment:
nit: you could test equals and hashcode using EqualsHashCodeTestUtils

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/91640/commits/c21262b78a3faa152e865a7e834ff858c4cc3f10.